### PR TITLE
Fix s3 crt checksum calc

### PR DESCRIPTION
**Pull request by sbiscigl**
*Description of changes:*

right now we don't propagate flexsum configuration to S3Crt, this changes to propagate the configuraiton. Previously the following request would fail.

```
#include <aws/core/Aws.h>
#include <aws/core/utils/HashingUtils.h>
#include <aws/s3-crt/S3CrtClient.h>
#include <aws/s3-crt/model/PutObjectRequest.h>

using namespace Aws;
using namespace Aws::Client;
using namespace Aws::S3Crt;
using namespace Aws::S3Crt::Model;

static auto ALLOC_TAG = "sdk-example";
static auto BUCKET_NAME = "bucket-name";
static auto KEY_NAME = "test-key";

auto main() -> int
{
    SDKOptions options;
    options.loggingOptions.logLevel = Utils::Logging::LogLevel::Debug;
    InitAPI(options);
    {
        S3CrtClient client;
        auto putObjectRequest = PutObjectRequest().WithBucket(BUCKET_NAME)
                                                  .WithKey(KEY_NAME);

        std::shared_ptr<IOStream> data = Aws::MakeShared<StringStream>(ALLOC_TAG,
            "test string",
            std::ios_base::in | std::ios_base::binary);

        putObjectRequest.SetBody(data);
        putObjectRequest.SetChecksumAlgorithm(ChecksumAlgorithm::CRC32);
        const auto putObjectResult = client.PutObject(putObjectRequest);
        assert(putObjectResult.IsSuccess());
    }
    ShutdownAPI(options);
    return 0;
}
```

Additionally this fixes memory leaks that existed in the CRT client that were not being tested before that now exist. This PR additionally re-enables the CRT integration tests so that we don't accidentally miss these issues anymore either.

*Check all that applies:*
- [x] Did a review by yourself.
- [x] Added proper tests to cover this PR. (If tests are not applicable, explain.)
- [x] Checked if this PR is a breaking (APIs have been changed) change.
- [x] Checked if this PR will _not_ introduce cross-platform inconsistent behavior.
- [x] Checked if this PR would require a ReadMe/Wiki update.

Check which platforms you have built SDK on to verify the correctness of this PR.
- [x] Linux
- [x] Windows
- [x] Android
- [x] MacOS
- [ ] IOS
- [ ] Other Platforms

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: generated/src/aws-cpp-sdk-s3-crt/source/S3CrtClient.cpp [hunk 1]
```diff
@@ -788,6 +788,33 @@ void S3CrtClient::CopyObjectAsync(const CopyObjectRequest& request, const CopyOb
...
+      {ChecksumAlgorithm::CRC32C, aws_s3_checksum_algorithm::AWS_SCA_CRC32C},
+      {ChecksumAlgorithm::SHA1, aws_s3_checksum_algorithm::AWS_SCA_SHA1},
+      {ChecksumAlgorithm::SHA256, aws_s3_checksum_algorithm::AWS_SCA_SHA256},
+    };
+
+    const auto checksumAlgorithm = request.GetChecksumAlgorithm();
+    const auto mapping = std::find_if(std::begin(crtChecksumMapping),
+      std::end(crtChecksumMapping),
+      [&checksumAlgorithm](const std::pair<ChecksumAlgorithm, aws_s3_checksum_algorithm>& mapping){ return mapping.first == checksumAlgorithm; });
+    if (mapping != std::end(crtChecksumMapping))
```
Comment (SergeyRyabinin) [inline generated/src/aws-cpp-sdk-s3-crt/source/S3CrtClient.cpp:808]: do we need to trace any message if mapped CRT algorithm hasn't been found?

**Review by SergeyRyabinin — APPROVED** [review]

File: generated/src/aws-cpp-sdk-s3-crt/source/S3CrtClient.cpp (hunk 1 above)
Comment (sbiscigl) [reply generated/src/aws-cpp-sdk-s3-crt/source/S3CrtClient.cpp:808]: added log

**Review by SergeyRyabinin — APPROVED** [review]

File: tests/aws-cpp-sdk-s3-crt-integration-tests/RunTestsWithMemTracer.cpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+int main(int argc, char** argv) {
+    Aws::Testing::SetDefaultSigPipeHandler();
+    Aws::SDKOptions options;
+    options.loggingOptions.logLevel = Aws::Utils::Logging::LogLevel::Trace;
+    CRTMemTracerMemorySystem memorySystem{};
+    options.memoryManagementOptions.memoryManager = &memorySystem;
+    Aws::Testing::InitPlatformTest(options);
+    Aws::Testing::ParseArgs(argc, argv);
+
+    Aws::InitAPI(options);
```
Comment (SergeyRyabinin) [inline tests/aws-cpp-sdk-s3-crt-integration-tests/RunTestsWithMemTracer.cpp:21]: I think Begin() goes here and End() right after "ShutdownAPI"

File: tests/testing-resources/include/aws/testing/MemoryTesting.h [hunk 3]
```diff
@@ -108,6 +110,19 @@ class AWS_TESTING_API ExactTestMemorySystem : public BaseTestMemorySystem
 
 };
 
+class AWS_TESTING_API CRTMemTracerMemorySystem : public Aws::Utils::Memory::MemorySystemInterface
+{
+public:
+    CRTMemTracerMemorySystem();
+    void Begin() override {}
+    void End() override {}
```
Comment (SergeyRyabinin) [inline tests/testing-resources/include/aws/testing/MemoryTesting.h:118]: I guess no allocations are allowed before Begin() and after End()

File: tests/aws-cpp-sdk-s3-crt-integration-tests/RunTestsWithMemTracer.cpp (hunk 2 above)
Comment (sbiscigl) [reply tests/aws-cpp-sdk-s3-crt-integration-tests/RunTestsWithMemTracer.cpp:21]: this is copy and pasted more or less from what we [already do](https://github.com/aws/aws-sdk-cpp/blob/main/tests/testing-resources/include/aws/testing/MemoryTesting.h#L131) and how we use it in the [crt test already](https://github.com/aws/aws-sdk-cpp/blob/main/tests/aws-cpp-sdk-s3-crt-integration-tests/RunTests.cpp#L17)

File: tests/testing-resources/include/aws/testing/MemoryTesting.h (hunk 3 above)
Comment (sbiscigl) [reply tests/testing-resources/include/aws/testing/MemoryTesting.h:118]: sorry not following whats needs to be done here? these are optional calls for static setup and we dont do any static setup so we can skip them

**Review by SergeyRyabinin — APPROVED** [review]

**Review by DmitriyMusatkin — APPROVED** [review]

**Pull request merged**